azure/services/natgateways/spec_test.go

```diff
@@ -0,0 +1,93 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package natgateways
+
+import (
+	"context"
+	"testing"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/network/armnetwork/v4"
+	. "github.com/onsi/gomega"
+	"k8s.io/utils/ptr"
+)
+
+var (
+	natGateway = armnetwork.NatGateway{
+		ID: ptr.To("/subscriptions/my-sub/resourceGroups/my-rg/providers/Microsoft.Network/natGateways/my-node-natgateway-1"),
+		Properties: &armnetwork.NatGatewayPropertiesFormat{
+			PublicIPAddresses: []*armnetwork.SubResource{
+				{ID: ptr.To("/subscriptions/my-sub/resourceGroups/my-rg/providers/Microsoft.Network/natGateways/pip-node-subnet")},
+			},
+		},
+	}
+)
+
+func TestNatGatewaySpec_Parameters(t *testing.T) {
+	testCases := []struct {
+		name          string
+		spec          *NatGatewaySpec
+		existing      interface{}
+		expect        func(g *WithT, result interface{})
+		expectedError string
+	}{
+		{
+			name:     "error when existing is not of NatGateway type",
+			spec:     &NatGatewaySpec{},
+			existing: struct{}{},
+			expect: func(g *WithT, result interface{}) {
+				g.Expect(result).To(BeNil())
+			},
+			expectedError: "struct {} is not an armnetwork.NatGateway",
+		},
+		{
+			name:     "get result as nil when existing NatGateway is present",
+			spec:     &natGatewaySpec1,
+			existing: natGateway,
+			expect: func(g *WithT, result interface{}) {
+				g.Expect(result).To(BeNil())
+			},
+			expectedError: "",
+		},
+		{
+			name:     "get NatGateway when all values are present",
+			spec:     &natGatewaySpec1,
+			existing: nil,
+			expect: func(g *WithT, result interface{}) {
+				g.Expect(result).To(BeAssignableToTypeOf(armnetwork.NatGateway{}))
+				g.Expect(result.(armnetwork.NatGateway).Location).To(Equal(ptr.To[string](natGatewaySpec1.Location)))
+				g.Expect(result.(armnetwork.NatGateway).Name).To(Equal(ptr.To[string](natGatewaySpec1.ResourceName())))
+			},
+			expectedError: "",
+		},
```

Comment:
What are your thoughts about adding a test case with empty NatGateway i.e  `&armnetwork.NatGateway{}`

Reply:
Added

Reply:
Here is the one with empty NatGateway (I have to define `Properties` as well otherwise it will fail with nil pointer): https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/4021/files#diff-4c03859149181d28bc73b23f73d171b0fbff4be635cbd24d9eb89e016734c138R77